### ISSUE TITLE
Optimize rendering, style recalculation and paint v1

### ISSUE THREAD

**Issue by BioPhoton** [BioPhoton/angular-movies-perf#78]
To optimize style recalculation layout and paint we can apply the `contain` styling rule

## TL;DR

- Content section: from 513ms to 251 ms (262ms improvement)
- Sidebar section: from 53ms to 32 ms (21ms improvement)
- Toolbar section: marginal animation improvement

**Total improvement 283ms**

## Movies section
When navigating to a movie list the whole document is rendered
![angular-movies-_movi-list-before](https://user-images.githubusercontent.com/10064416/131185073-c1d7a475-24e4-456d-bc4b-3e72de2e4aac.PNG)

```html
<div _ngcontent-hub-movies-c24="" class="content">
 <!-- movie list -->
</div>
```

the stylings for the content section are:
```css
.content[_ngcontent-hub-movies-c24] {
    min-height: calc(100vh - 72px);
    width: 100%;
}
```

**Measures Before:**
![angular-movies-_movi-list-detail-before](https://user-images.githubusercontent.com/10064416/131185990-ab8396e3-9849-40ac-83c8-894adad2b656.PNG)

- 0 ms  Scripting
- 415 ms  Rendering
- 56 ms  Painting
- 40 ms  System
- 1 ms  Idle
- **513 ms  Total**

To improve the render cost we can apply some styling rules.

```css
.content[_ngcontent-hub-movies-c24] {
  ...
  contain: strict;
}
```

This scopes the area that gets rerendered to the div.
![angular-movies-_movi-list-after](https://user-images.githubusercontent.com/10064416/131186048-769b48a8-afa6-4009-9484-8c18f6e4e7e1.PNG)

**Measures After:**
![angular-movies-_movi-list-detail-after](https://user-images.githubusercontent.com/10064416/131186058-2a77237e-5b8e-417d-85fd-f183818b98c7.PNG)

- 1 ms  Scripting
- 232 ms  Rendering
- 11 ms  Painting
- 8 ms  System
- 0 ms  Idle
- **251 ms  Total**

**The overall improvement is minus 262ms** 😎

## Side drawer 

The side drawer gets an update of fetched menu items. These causes rerender of the whole page. 

Again we apply `contain: strict;`.

```html
<div class="side-drawer">
 <!-- menu items -->
</div>
```

**Measures Before:**
![angular-movies-_drawer-detail-before](https://user-images.githubusercontent.com/10064416/131191650-f56428fb-1114-486d-89bd-076ed62dd4d5.PNG)
![angular-movies-_drawer-before](https://user-images.githubusercontent.com/10064416/131191665-ad2e91f8-12a7-4c4a-bd89-a9104f76d34d.PNG)

- 0 ms  Scripting
- 3 ms  System (self)
- 44 ms  Rendering
- 6 ms  Painting
- 0 ms  Idle
- **53 ms  Total**

**Measures After:**
![angular-movies-_drawer-detail-after](https://user-images.githubusercontent.com/10064416/131191680-4d10e305-dc3c-4837-8147-7dea5cd0c486.PNG)
![angular-movies-_drawer-after](https://user-images.githubusercontent.com/10064416/131191683-b6000839-b98a-4b95-8e19-ede86c7bad4f.PNG)

- 0 ms  Scripting
- 25 ms  Rendering
- 4 ms  Painting
- 2 ms  System
- 0 ms  Idle
- **32 ms  Total**

**The overall improvement is minus 31ms** 🙂

## Toolbar and input animation

This is a small one but the approach is the same.

Here we apply `contain: strict;` to the app-toolbar component:

```html
<div _ngcontent-hub-movies-c24="" class="app-toolbar">
 <!-- header toolbar -->
</div>
```

**Measures Before:**
![angular-movies-_toolbar-before](https://user-images.githubusercontent.com/10064416/131187809-a42d8ed3-289b-45da-81a4-90eda90e5a77.PNG)

**Measures After:**

![angular-movies-_toolbar-after](https://user-images.githubusercontent.com/10064416/131187871-fda20280-df5a-40b6-8748-dc4f2fa3001e.PNG)

The impact is marginal but still fewer items processed.
